### DISCUSSION TITLE
Blocks getting deleted randomly

### DISCUSSION THREAD

**Discussion by ghost** [area17/twill#1428] [Q&A]
Hi,

Every so often, records in the blocks table get deleted when they shouldnt. This happens sporadically and is difficult to replicate, but would there be any reason why the blocks table would get truncated or records would be removed? Maybe something to do with Revisions?

Sorry I can't give much more info.

Cheers,
Aaron

**Comment by ifox** [comment]
Hi @aarontilt, what do you mean by "when they shouldn't"? If you mean when saving even though they have not changed, that is expected, but if you mean something else, that's not something we've experienced.

**Reply by ghost** [reply]
Hi,

Cheers for the response. Essentially we get intermittent issues where a user will create a bunch of blocks attached to an object, then will save it, and everything will be fine for a while. Then occasionally the blocks will disappear from the database, seemingly without touching them or any related record. Any ideas what's going on?

**Reply by ifox** [reply]
Unfortunately, I really do not, as there is no code within Twill that would delete blocks outside of a user saving the form. Are you relying on the blocks IDs somehow? If that's the case, that would be your issue, since even though they are not changed, simply saving the form again does delete and re-create them under new IDs in the `blocks` table.

**Reply by ghost** [reply]
I dont believe we're relying on block IDs, so I don't think that's the issue. Cheers anyway, I'll do some more digging

**Reply by ghost** [reply]
UPDATE: So, we've discovered the cause, and I'm wondering if you have an idea of what the solution could be:

When moving an item around in a repeater, if the items in the repeater have blocks related to them, all blocks related to the items in the repeater will get deleted. e.g.
- I have a class called Models
- A Model can be assigned any number of blocks
- I have another class called ModelList
- ModelList has a repeater that contains a list of Models
- If I reorder the Models list using the ModelList repeater and update the ModelList, the blocks associated with all Models in the repeater are deleted.

Any thoughts?

**Reply by ifox** [reply]
Wow ok, I totally get what's going on now! You're using both the "repeater" form and the "form" form of your Model form (pun intended!). Correct? So what's happening is that when you save the form that contains the repeater, its blocks are not there, so it deletes them, as its not aware they are not there because you don't want them to be edited from the repeater itself.

I don't think anyone in the community has been using repeaters in that capacity, or if they have, they used the same form fields on both sides. What most would do in this case is actually use a browser field and not mess with the ability to edit the same thing in multiple places. 

Now, I totally get the use case and do want to make it work. I actually believe the work we are doing on #1427 should help fix it.

**Reply by haringsrob** [reply]
Just pointing to the correct pr regarding this: https://github.com/area17/twill/pull/1431

**Reply by ghost** [reply]
Hi,

Thanks for the update! It looks like you've got something in the works for this which is good to hear.

Just to clarify my scenario:
- I have a class called `Model`
- A `Model` can be given any number of `Blocks` via a block_editor field
- I have another class called `ModelList`
- A `ModelList` can be given any number of `Models` via a repeater field

So the relation chain goes `ModelList` 1:M `Model` 1:M `Block`

Going to the `ModelList` record in the CMS and reordering the `Models` repeater will cause all `Blocks` associated with all `Models` in the `ModelList` repeater to be deleted. I'm assuming this is cuz stuff gets deleted and remade as a result of the reorder?

Hopefully that doesn't change your fix!
